### PR TITLE
Provide a means to preserve session order.

### PR DESCRIPTION
**Pull request by lukesneeringer**
Okay @jonparrott, I am a convert to nox. It is better. Mostly. ;-)

One thing that bothers me quite a bit (like @tswast) is that nox does not honor the order that sessions are declared in `nox.py`, instead running your sessions in alphabetical order. I really want unit tests to run before system tests, for example, and also I want my unit tests to amalgamate coverage on each interpreter and spit out a unified report at the end.

The current implementation relies on reading `module.__dict__`, which is unordered, so I am proposing an API change. I have added a decorator, `@nox.session`, which causes nox to recognize a function as a session. This is intended to be the preferred approach instead of relying on the convention of naming functions `session_*`, although the previous functionality remains in place for backwards compatibility.

If accepted, this will work as follows:

  * The preferred way to declare a nox session is with `@nox.session`. The function name is retained as-is.
  * All sessions that use the decorator are run in order. Sessions that use the naming convention still run alphabetically. If you mix and match, all the decorator sessions run first (in order), and then the naming convention sessions (alphabetically).

This has one side effect:

  * The `nox.session` module is renamed `nox.sessions`, to avoid a naming collision. This is ostensibly a backwards incompatible change, but I am fairly certain the old module was unlikely to be directly used by a downstream user.

Also worth noting, I designed `SessionRegistry` to be a little more robust than what the current feature set needs. I could see several additional uses for it in the future. Two obvious examples I can think of:

  * "run everything with an interpreter of Python 3.6"
  * declaring registries with explicit names and then running all sessions in a given registry.

Feedback welcome. :-)

### CODE REVIEW

**Comment by lukesneeringer** [comment]
Additionally: Fixes #11.

**Comment by lukesneeringer** [comment]
Sadly I have no idea why the AppVeyor build is failing, and the "environment can only contain strings" error message is not helpful...

**Comment by jonparrott** [comment]
Hmm.. doesn't pytest do discovery and preserve order? I wonder how they pull that off.

> Also worth noting, I designed SessionRegistry to be a little more robust than what the current feature set needs. I could see several additional uses for it in the future. Two obvious examples I can think of:

Yeah this is super overkill. Can we simplify it for now (YAGNI)? We'll have the more robust one in the git history if we need it in the future, and we won't have to change the interface.

**Comment by lukesneeringer** [comment]
> Yeah this is super overkill. Can we simplify it for now (YAGNI)? We'll have the more robust one in the git history if we need it in the future, and we won't have to change the interface.

Sure, although you will have to merge without squash to have it in git history. :-)

**Comment by jonparrott** [comment]
I'm happy to do that, just do an interactive rebase so that we have pretty
history. :)

**Comment by lukesneeringer** [comment]
As long as I make the new one a single commit, we will have a pretty history. Each of the two commits there now do a distinct thing.

**Comment by lukesneeringer** [comment]
@jonparrott The simplified registry is in. I do ask that you preserve the history in the second commit if possible (being able to "group" sessions into a registry and run a registry is something that I know I am going to want; it would be great for google-cloud-python).

**Comment by jonparrott** [comment]
> being able to "group" sessions into a registry and run a registry is something that I know I am going to want; it would be great for google-cloud-python.

I'd like to take (another) page out of pytest's book and use `-k 'something'` to do matching and add `nox.mark`.

File: setup.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
 setup(
     name='nox-automation',
 
-    version='0.9.1',
+    version='0.9.2',
```
Comment (jonparrott) [inline setup.py]: Go ahead and bump to `0.10.0`

**Comment by lukesneeringer** [comment]
> I'd like to take (another) page out of pytest's book and use -k 'something' to do matching and add `nox.mark`.

Got it. I might implement that next week.

**Comment by jonparrott** [comment]
> Got it. I might implement that next week.

Awesome. :)

**Comment by jonparrott** [comment]
I was curious how Sphinx autodoc can preserve member order in the docs. Turns out it analyzes the source code to do so (see [here](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/pycode/__init__.py#L291)). I would love something like that here so we can continue to look like `py.test`, but alas, it's complicated.

**Comment by lukesneeringer** [comment]
One nice thing is that even if we implement this functionality now, we could still implement that later. It would require that you continue to support both syntaxes, but I do not see that as a huge problem. (I might be biased, though, because I actually like the decorator syntax -- "explicit is better than implicit".)

**Comment by jonparrott** [comment]
> I actually like the decorator syntax -- "explicit is better than implicit".)

I do too, I just want to try to remain as close to `py.test` as possible.

> is there anything else this requires?

No, I'm just working on perf. I'll give it another once-over and merge soon.

**Comment by jonparrott** [comment]
@lukesneeringer this is released. :)

**Comment by lukesneeringer** [comment]
Cheers! Thanks @jonparrott.

If time permits, I might do a couple of the other small features discussed in this issue later this week.

**Comment by jonparrott** [comment]
> If time permits, I might do a couple of the other small features discussed in this issue later this week.

That would be awesome.

**Pull request merged**